File: medium/encode_strings.py
# Encode and Decode Strings - Medium
# 
# Problem Description:
# Design an algorithm to encode a list of strings to a string and decode
# the string back to the original list of strings.
# 
# Example 1:
# Input: strs = ["hello","world"]
# Output: "5#hello5#world"
# Decoded: ["hello","world"]
# 
# Example 2:
# Input: strs = [""]
# Output: "0#"
# Decoded: [""]
#
# Constraints:
# 0 <= strs.length <= 200
# 0 <= strs[i].length <= 200
# strs[i] contains any possible characters out of 256 valid ASCII characters.
#
# Approach (Length Prefixing):
# Encode: For each string, prepend its length followed by '#' delimiter
# Decode: Parse length, skip delimiter, extract exact number of characters
#
# Time Complexity: O(n) - where n is total characters across all strings
# Space Complexity: O(n) - for the encoded string and result list

from typing import List

class Solution:
    
    def encode(self, strs: List[str]) -> str:
        res = ""
        for string in strs:
            # Format: length + delimiter + actual string
            res += str(len(string)) + "#" + string
        return res
    
    def decode(self, s: str) -> List[str]:
        i = 0
        res = []
        
        while i < len(s):
            # Find the '#' delimiter from current position
            indexHash = s.find("#", i)
            # Extract length (between current position and '#')
            length = int(s[i:indexHash])
            # Start of actual string is after the '#'
            start = indexHash + 1
            # Extract exact number of characters based on length
            res.append(s[start:start+length])
            # Move to start of next encoded string
            i = length + start
        
        return res

File: medium/encode_strings.js
// Encode and Decode Strings - Medium
//
// Problem Description:
// Design an algorithm to encode a list of strings into a single string and decode
// the string back into the original list of strings.
//
// Example 1:
// Input: strs = ["hello", "world"]
// Encoded Output: "5#hello5#world"
// Decoded Output: ["hello", "world"]
//
// Example 2:
// Input: strs = [""]
// Encoded Output: "0#"
// Decoded Output: [""]
//
// Constraints:
// - 0 <= strs.length <= 200
// - 0 <= strs[i].length <= 200
// - strs[i] can contain any valid ASCII characters.
//
// Approach (Length Prefixing):
// Encode: For each string, prepend its length followed by a '#' delimiter.
// Decode: Parse the length, skip the delimiter, and extract the exact number of characters.
//
// Time Complexity: O(n) - where n is the total number of characters across all strings.
// Space Complexity: O(n) - for the encoded string and the result list.

class Solution {
    /**
     * Encodes a list of strings into a single string.
     * @param {string[]} strs - The list of strings to encode.
     * @returns {string} - The encoded string.
     */
    encode(strs) {
        let res = "";
        for (const str of strs) {
            res += String(str.length) + "#" + str;
        }
        return res;
    }

    /**
     * Decodes a single string back into a list of strings.
     * @param {string} str - The encoded string.
     * @returns {string[]} - The decoded list of strings.
     */
    decode(str) {
        const res = [];
        let i = 0;

        while (i < str.length) {
            let hashIndex = str.indexOf("#", i);
            if (hashIndex === -1) break;

            let length = parseInt(str.substring(i, hashIndex));

            let start = hashIndex + 1;
            let end = start + length;
            i = end;
            res.push(str.substring(start, end));
        }
        return res;
    }
}